mlflow/sagemaker/__init__.py

```diff
@@ -316,36 +388,121 @@ def _upload_s3(local_model_path, bucket, prefix):
+        
+        raise Exception(msg)
+    
+    elif endpoint_found:
+        return _update_sagemaker_endpoint(endpoint_name=app_name,
+                                          image_url=image_url,
+                                          model_s3_path=model_s3_path,
+                                          run_id=run_id,
+                                          mode=mode, 
+                                          archive=archive,
+                                          role=role,
+                                          sage_client=sage_client,
+                                          s3_client=s3_client)
+
+    else:
+        return _create_sagemaker_endpoint(endpoint_name=app_name,
+                                          image_url=image_url,
+                                          model_s3_path=model_s3_path,
+                                          run_id=run_id,
+                                          role=role,
+                                          sage_client=sage_client)
+
+
+def _get_sagemaker_resource_unique_id():
+    """
+    :return: A unique, identifier that can be appended to a user-readable 
+    resource name to avoid naming collisions.
+    """
+    uuid_bytes = uuid.uuid4().bytes
+    # Use base64 encoding to shorten the UUID length. Note that the replacement of the 
+    # unsupported '+' symbol maintains uniqueness because the UUID byte string is of a fixed,
+    # 32-byte length
+    uuid_b64 = uuid_bytes.encode("base64").rstrip('=\n').replace("/", "-").replace("+", "AB")
+    return uuid_b64
+
+def _get_sagemaker_model_name(endpoint_name):
+    unique_id = _get_sagemaker_resource_unique_id()
+    return "{en}-model-{uid}".format(en=endpoint_name,
+                                     uid=unique_id)
+
+def _get_sagemaker_config_name(endpoint_name):
+    unique_id = _get_sagemaker_resource_unique_id()
+    return "{en}-config-{uid}".format(en=endpoint_name,
+                                      uid=unique_id)
+
+def _create_sagemaker_endpoint(endpoint_name, 
+                               image_url, 
+                               model_s3_path, 
+                               run_id, 
+                               role, 
+                               sage_client): 
+    """
+    :param image_url: URL of the ECR-hosted docker image the model is being deployed into
+    :param model_s3_path: s3 path where we stored the model artifacts
+    :param run_id: RunId that generated this model
+    :param role: SageMaker execution ARN role
+    :param sage_client: A boto3 client for SageMaker
+    """
+    eprint("Creating new endpoint with name: {en} ...".format(en=endpoint_name))
+
+    model_name = _get_sagemaker_model_name(endpoint_name)
     model_response = sage_client.create_model(
         ModelName=model_name,
         PrimaryContainer={
-            'ContainerHostname': 'mlflow-serve-%s' % model_name,
+            'ContainerHostname': 'mfs-serve-%s' % model_name,
```

Comment:
?

Reply:
Whoops. I meant to replace `mlflow-serve` with `mfs` because I was frequently hitting the 63-character limit for  SageMaker resource names. I've now made that adjustment. Please let me know if you disagree with this approach.